controllers/nim_account_controller.go

```diff
@@ -0,0 +1,480 @@
+		Owns(&corev1.ConfigMap{}).
+		Owns(&corev1.Secret{}).
+		Owns(&templatev1.Template{}).
+		Watches(&corev1.Secret{}, handler.EnqueueRequestsFromMapFunc(
+			func(ctx context.Context, obj client.Object) []reconcile.Request {
+				var requests []reconcile.Request
+				accounts := &v1.AccountList{}
+				if err := mgr.GetClient().List(ctx, accounts, client.MatchingFields{apiKeySpecPath: obj.GetName()}); err != nil {
+					r.Log.Error(err, "failed to fetch accounts")
+					return requests
+				}
+				for _, item := range accounts.Items {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{
+						Name:      item.Name,
+						Namespace: item.Namespace,
+					}})
+				}
+				return requests
+			})).
+		Complete(r)
+}
+
+func (r *NimAccountReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Log.WithValues("Account", req.Name, "namespace", req.Namespace)
+	ctx = log.IntoContext(ctx, logger)
+
+	account := &v1.Account{}
+	if err := r.Client.Get(ctx, req.NamespacedName, account); err != nil {
+		if k8serrors.IsNotFound(err) {
+			logger.V(1).Info("account deleted")
+		} else {
+			logger.V(1).Error(err, "failed to fetch object")
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !account.DeletionTimestamp.IsZero() {
+		logger.V(1).Info("account being deleted")
+		return ctrl.Result{}, nil
+	}
+
+	logger.V(1).Info("account active")
+
+	initialMsg := "not reconciled yet"
+	targetStatus := &v1.AccountStatus{
+		Conditions: []metav1.Condition{
+			// initial unknown status
+			makeAccountSuccessfulCondition(account.Generation, initialMsg),
+			makeApiKeyUnknownCondition(account.Generation, initialMsg),
+			makeConfigMapUnknownCondition(account.Generation, initialMsg),
+			makeTemplateUnknownCondition(account.Generation, initialMsg),
+			makePullSecretUnknownCondition(account.Generation, initialMsg),
+		},
+	}
+
+	defer func() {
+		r.updateStatus(ctx, req.NamespacedName, *targetStatus)
+	}()
+
+	// fetch api secret
+	secretNs := account.Spec.APIKeySecret.Namespace
+	if secretNs == "" {
+		secretNs = account.Namespace
+	}
+	apiKeySecret := &corev1.Secret{}
+	apiKeySecretSubject := types.NamespacedName{Name: account.Spec.APIKeySecret.Name, Namespace: secretNs}
+	if err := r.Client.Get(ctx, apiKeySecretSubject, apiKeySecret); err != nil {
+		var msg string
+		if k8serrors.IsNotFound(err) {
+			msg = "api key secret was deleted"
+			logger.Info(msg)
+		} else {
+			msg = "failed to fetch api key secret"
+			logger.V(1).Error(err, "failed to fetch api key secret")
+		}
+		meta.SetStatusCondition(&targetStatus.Conditions, makeAccountFailureCondition(account.Generation, msg))
+		r.cleanupResources(ctx, account)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	foundApiKeySec := "found api key secret"
+	logger.V(1).Info(foundApiKeySec)
+	meta.SetStatusCondition(&targetStatus.Conditions, makeAccountFailureCondition(account.Generation, foundApiKeySec))
```

Comment:
nit: Consider making an util function to log + set condition since these are repeated many times

Reply:
Also, while trying this, I noticed I missed an unknown condition. Resolved here https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/b8293f4d86d0821c78a8b52c56a7720f8ed4228f.

Reply:
Let me know if you want me to do the utility funcs regardless.